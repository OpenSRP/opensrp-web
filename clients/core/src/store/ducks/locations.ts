import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'locations';

/** A single node for location hierarchy */
export interface LocationNode {
    [key: string]: {
        id: string;
        label: string;
        node: {
            locationId: string;
            name: string;
            parentLocation: {
                locationId: string;
                name: string;
                voided: boolean;
                parentLocation?: {
                    locationId: string;
                    name: string;
                    voided: boolean;
                };
            };
            tags: string[];
            voided: boolean;
        };
        children?: LocationNode;
    };
}

/** Interface for location object */
export interface LocationHierarchy {
    locationsHierarchy: {
        map: LocationNode;
    };
}

/** SET_LOCATION action type */
export const SET_LOCATIONS = 'opensrp/reducer/locations/SET_LOCATIONS';

/** Interface for setLocationAction */
interface SetLocationAction extends AnyAction {
    locations: LocationHierarchy;
    type: typeof SET_LOCATIONS;
}

/** Create type for location reducer actions */
export type LocationActionTypes = SetLocationAction | AnyAction;

/** setLocation action */
export const setLocations = (locations: LocationHierarchy): SetLocationAction => ({
    locations,
    type: SET_LOCATIONS,
});
