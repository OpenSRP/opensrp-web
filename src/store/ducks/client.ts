import { AnyAction } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import { Client } from './clients';

/** Interface for event object as received from event search */
export interface Event {
  type: 'Event';
  dateCreated: number;
  serverVersion: number;
  clientApplicationVersion: number;
  clientDatabaseVersion: number;
  identifiers: { [key: string]: string | null };
  baseEntityId: string;
  locationId: string;
  eventDate: number;
  eventType: string;
  formSubmissionId: string;
  providerId: string;
  duration: number;
  obs: { [key: string]: string[] | string | null };
  entityType: string;
  version: number;
  teamId: string;
  team: string;
  _id: string;
  _rev: string;
}

/** The reducer name */
export const reducerName = 'client';

/** CLIENT_FETCHED action type */
export const CLIENT_FETCHED = 'opensrp/reducer/client/CLIENT_FETCHED';

/** interface for fetch client */
export interface FetchClientAction extends AnyAction {
  clientById: { [key: string]: Client };
  type: typeof CLIENT_FETCHED;
}

// actions

// action Creators

/** Fetch client action creator
 * @param {Client} client - client to add to store
 * @return {FetchClientAction} - an action to add client to redux store
 */
export const fetchClient = (client: Client): FetchClientAction => ({
  clientById: { [client.baseEntityId]: client },
  type: CLIENT_FETCHED,
});

/** Create type for client reducer actions */
export type ClientActionTypes = FetchClientAction | AnyAction;

// The reducer

/** interface for client state in redux store */
interface ClientState {
  clientById: { [key: string]: Client };
  eventsById: { [key: string]: Event };
  membersById: { [key: string]: Client };
}

/** Create an immutable client state */
export type ImmutableClientState = ClientState & SeamlessImmutable.ImmutableObject<ClientState>;

/** initial clients-state state */
const initialState: ImmutableClientState = SeamlessImmutable({
  clientById: {},
  eventsById: {},
  membersById: {},
});

/** the client reducer function */
export default function reducer(
  state: ImmutableClientState = initialState,
  action: ClientActionTypes
): ImmutableClientState {
  switch (action.type) {
    case CLIENT_FETCHED:
      return SeamlessImmutable({
        ...state,
        clientById: action.client,
      });
    default:
      return state;
  }
}
