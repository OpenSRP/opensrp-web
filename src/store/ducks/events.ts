import { AnyAction, Store } from 'redux';
import { keyBy, values } from 'lodash';
import SeamlessImmutable from 'seamless-immutable';

/** The reducer name */
export const reducerName = 'events';

/** Interface for event object as received from event search */
export interface Event {
    type: 'Event';
    dateCreated: number;
    serverVersion: number;
    clientApplicationVersion: number;
    clientDatabaseVersion: number;
    identifiers: { [key: string]: string | null };
    baseEntityId: string;
    locationId: string;
    eventDate: number;
    eventType: string;
    formSubmissionId: string;
    providerId: string;
    duration: number;
    obs: { [key: string]: string[] | string | null };
    entityType: string;
    version: number;
    teamId: string;
    team: string;
    _id: string;
    _rev: string;
  }

  /** EVENTS_FETCHED action type */
export const EVENTS_FETCHED = 'opensrp/reducer/client/EVENTS_FETCHED';

/** interface for fetch events */
export interface FetchEventsAction extends AnyAction {
    eventsById: { [key: string]: Event };
    type: typeof EVENTS_FETCHED;
}

/** Fetch events action creator
 * @param {Event} events - events to add to store
 * @return {FetchEventsAction} - an action to add events to redux store
 */
export const fetchEvents = (eventsList: Event[]): FetchEventsAction => ({
    eventsById: keyBy(eventsList, (event: Event) => event.baseEntityId),
    type: EVENTS_FETCHED,
});

/** Create type for client reducer actions */
export type EventActionTypes = FetchEventsAction | AnyAction;
  
/** interface for client state in redux store */
interface EventState {
    eventsById: { [key: string]: Event };
}

/** Create an immutable client state */
export type ImmutableClientState = EventState & SeamlessImmutable.ImmutableObject<EventState>;

/** initial clients-state state */
const initialState: ImmutableClientState = SeamlessImmutable({
    eventsById: {},
});

/** the client reducer function */
export default function reducer(
    state: ImmutableClientState = initialState,
    action: EventActionTypes
  ): ImmutableClientState {
    switch (action.type) {
      case EVENTS_FETCHED:
        return SeamlessImmutable({
          ...state,
          eventsById: action.eventsById,
        });
      
      default:
        return state;
    }
  }

/** returns the events of the client in the store as array
 * @param {Partial<Store>} state - the redux store
 * @return { Event } - events array
 */
export function getEventsArray(state: Partial<Store>): Event[] {
    return values((state as any)[reducerName].eventsById);
  }