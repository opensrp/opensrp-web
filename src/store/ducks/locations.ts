import { LocationState } from 'history';
import { keyBy, values } from 'lodash';
import { AnyAction, Store } from 'redux';
import { COMMUNE, DISTRICT, PROVINCE, VILLAGE } from '../../constants';

/** the reducer name */
export const reducerName = 'LocationReducer';

/** Interface for Location object received from discover */
export interface Location {
  location_id: string;
  location_name: string;
  level: VILLAGE | COMMUNE | DISTRICT | PROVINCE;
  parent_id: string;
}

export interface UserLocation {
  provider_name: string;
  provider_contact: string;
  provider_id: string;
  location_id: string | any;
  phone_number?: string;
  location_name: string;
}

export interface Locations {
  [key: string]: Location | UserLocation;
}

// actions

/** FETCH_USER_LOCATION action type */
export const FETCHED_USER_LOCATION = 'opensrp/reducer/FETCH_USER_LOCATION';

/** FETCH_LOCATION action type */
export const FETCHED_LOCATION = 'opensrp/reducer/FETCH_LOCATION';

/** REMOVE_LOCATION action type */
export const REMOVE_LOCATIONS = 'opensrp/reducer/REMOVE_LOCATIONS';

/** Interface for FetchLocationAction */
export interface FetchLocationsAction extends AnyAction {
  locations: { [key: string]: Location };
  type: typeof FETCHED_LOCATION;
}

/** Interface for FetchUserLocationAction */
export interface FetchUserLocationsAction extends AnyAction {
  userLocations: { [key: string]: UserLocation };
  type: typeof FETCHED_USER_LOCATION;
}

export const removeLocations = {
  locations: [],
  type: REMOVE_LOCATIONS,
};

/** Location action types */
export type LocationActionTypes = FetchLocationsAction | FetchUserLocationsAction | AnyAction;

// action creators

/**
 * Fetch Location action creator
 * @param {Location[]} - Location array to add to store
 * @return {FetchLocationAction} - an action to add Location array to store
 */
export const fetchLocations = (locations: Location[] = []): FetchLocationsAction => {
  return {
    locations: keyBy(locations, (location: Location) => location.location_id),
    type: FETCHED_LOCATION as typeof FETCHED_LOCATION,
  };
};

export const fetchUserLocations = (
  userLocations: UserLocation[] = []
): FetchUserLocationsAction => {
  const fetchUserLocationsAction = {
    type: FETCHED_USER_LOCATION as typeof FETCHED_USER_LOCATION,
    userLocations: keyBy(userLocations, (userLocation: UserLocation) => userLocation.provider_id),
  };
  return fetchUserLocationsAction;
};

/** interface for locations state in redux store */
interface LocationsState {
  locations: { [key: string]: Location };
  locationsFetched: boolean;
  userLocations: { [key: string]: UserLocation };
  userLocationsFetched: boolean;
}

/** Initial location-state state */
const initialState: LocationsState = {
  locations: {},
  locationsFetched: false,
  userLocations: {},
  userLocationsFetched: false,
};

export default function locationsReducer(
  state: LocationsState = initialState,
  action: LocationActionTypes
): LocationState {
  switch (action.type) {
    case FETCHED_LOCATION:
      return {
        ...state,
        locations: { ...state.locations, ...action.locations },
        locationsFetched: true,
      };
    case FETCHED_USER_LOCATION:
      return {
        ...state,
        userLocations: { ...state.userLocations, ...action.userLocations },
        userLocationsFetched: true,
      };
    case REMOVE_LOCATIONS:
      return {
        ...state,
        locations: action.locations,
        locationsFetched: false,
      };
    default:
      return state;
  }
}

// Selectors

/** Return all Location data in the store as values whose
 * keys are their respective ids
 * @param {Partial<store>} state - the redux store
 * @return {Location[]}} - an array of Location objects
 */
export function getLocations(state: Partial<Store>): Location[] {
  return values((state as any)[reducerName].locations);
}

/** Return all User Location data in the store as values whose
 * keys are their respective provider ids
 * @param {Partial<store>} state - the redux store
 * @return {UserLocation[]}} - an array of User Location objects
 */
export function getUserLocations(state: Partial<Store>): UserLocation[] {
  return values((state as any)[reducerName].userLocations);
}

/**
 * Return locations of a certain level/administrative unit
 * @param {Partial<store>} state - the redux store
 * @param {String} level - the level/aministrative unit we would like to get from the store
 * @return {Location[]} - an array of location objects
 */
export function getLocationsOfLevel(state: Partial<Store>, level: string): Location[] {
  return values((state as any)[reducerName].locations).filter((location: Location) => {
    return location.level === level;
  });
}

/** Returns true if location data has been fetched from superset
 * and false otherwise
 */
export function locationsDataFetched(state: Partial<Store>): boolean {
  return (state as any)[reducerName].locationsFetched;
}

/** Returns true if user location details has been fetched from superset
 * and false otherwise
 */

export function userLocationDataFetched(state: Partial<Store>): boolean {
  return (state as any)[reducerName].userLocationsFetched;
}
