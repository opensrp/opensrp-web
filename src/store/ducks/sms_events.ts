import { get, keyBy, keys, values } from 'lodash';
import { AnyAction, Store } from 'redux';

/** The reducer name */
export const reducerName = 'SmsReducer';

/** Interface for SMS record object as received from discover */
// tslint:disable-next-line: class-name
export interface SmsData {
  age: string;
  EventDate: string;
  event_id: string;
  health_worker_location_name: string;
  message: string | number;
  anc_id: string;
  logface_risk: string;
  health_worker_name: string;
  sms_type: string;
  height: number;
  weight: number;
  previous_risks: string;
  lmp_edd: number;
  parity: number;
  gravidity: number;
  location_id: string;
}

// actions

/** fETCH_SMS action type */
export const FETCHED_SMS = 'opensrp/reducer/FETCHED_SMS';
/** REMOVE_SMS action type */
export const REMOVE_SMS = 'opensrp/reducer/REMOVE_SMS';

/** interface for sms fetch */
// tslint:disable-next-line: class-name
export interface FetchSmsAction extends AnyAction {
  smsData: { [key: string]: SmsData };
  type: typeof FETCHED_SMS;
}

/** Create type for SMS reducer actions */
export type SmsActionTypes = FetchSmsAction | AnyAction;

// action Creators

/** Fetch SMS action creator
 * @param {SmsData[]} smsData - SmsData array to add to store
 * @return {FetchSmsAction} - an action to add SmsData to redux store
 */
export const fetchSms = (smsDataList: SmsData[] = []): FetchSmsAction => {
  const actionCreated = {
    smsData: keyBy(smsDataList, (smsData: SmsData) => smsData.event_id),
    type: FETCHED_SMS as typeof FETCHED_SMS,
  };
  return actionCreated;
};

export const removeSms = {
  smsDataById: {},
  type: REMOVE_SMS,
};
// The reducer

/** interface for sms state in redux store */
// tslint:disable-next-line: class-name
interface SmsState {
  smsData: { [key: string]: SmsData };
  smsDataFetched: boolean;
}

/** initial sms-state state */
const initialState: SmsState = {
  smsData: {},
  smsDataFetched: false,
};

/** the sms reducer function */
export default function reducer(state: SmsState = initialState, action: SmsActionTypes): SmsState {
  switch (action.type) {
    case FETCHED_SMS:
      return {
        ...state,
        smsData: { ...state.smsData, ...action.smsData },
        smsDataFetched: true,
      };
    case REMOVE_SMS:
      return {
        ...state,
        smsData: action.smsData,
        smsDataFetched: false,
      };
    default:
      return state;
  }
}

// // Selectors

/** returns all data in the store as values whose keys are their respective ids
 * @param {Partial<Store>} state - the redux store
 * @return { { [key: string] : SmsData}[] } - SmsData object[] as values, respective ids as keys
 */
export function getSmsData(state: Partial<Store>): SmsData[] {
  return values((state as any)[reducerName].smsData);
}

/** returns true if sms data has been fetched from superset and false
 * if the data has not been fetched
 */
export function smsDataFetched(state: Partial<Store>): boolean {
  return (state as any)[reducerName].smsDataFetched;
}
