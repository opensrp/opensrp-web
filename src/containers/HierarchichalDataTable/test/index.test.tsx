import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createBrowserHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import ConnectedHierarchichalDataTable from '..';
import { fetchLocations } from '../../../store/ducks/locations';
import { fetchSms } from '../../../store/ducks/sms_events';
import store from '../../../store/index';
import { smsSlice } from '../../LogFace/tests/fixtures';
import { communes, districts, provinces, villages } from './fixtures';

jest.genMockFromModule('@fortawesome/react-fontawesome');
jest.mock('@fortawesome/react-fontawesome');

const history = createBrowserHistory();

describe('HierarchichalDataTable', () => {
  it('renders without crashing', () => {
    shallow(
      <Provider store={store}>
        <ConnectedHierarchichalDataTable />
      </Provider>
    );
  });

  it('renders correctly', () => {
    store.dispatch(fetchLocations(districts));
    store.dispatch(fetchLocations(provinces));
    store.dispatch(fetchLocations(communes));
    store.dispatch(fetchLocations(villages));
    store.dispatch(fetchSms(smsSlice));

    let props = {
      match: {
        params: {
          risk_highlighter: 'high_risk',
          title: 'Total Pregnancies',
        },
      },
    };

    let wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedHierarchichalDataTable {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('Table'))).toMatchSnapshot('hierarchichalDataTable');

    props = {
      match: {
        params: {
          current_level: 1,
          direction: 'down',
          node_id: '1',
          risk_highlighter: 'high_risk',
          title: 'Total Pregnancies',
        } as any,
      },
    };

    // pass props in the partern /:risk_highlighter?/:title?/:current_level?/:direction?/:node_id?/:from_level?
    //             hierarchicaldata/high-risk/156 Total Pregnancies/1/down/1
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedHierarchichalDataTable {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('tbody'))).toMatchSnapshot('hierarchichalDataTable');

    props = {
      match: {
        params: {
          current_level: 2,
          direction: 'down',
          node_id: '6',
          risk_highlighter: 'high_risk',
          title: 'Total Pregnancies',
        } as any,
      },
    };
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedHierarchichalDataTable {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('tbody'))).toMatchSnapshot('hierarchichalDataTable');

    props = {
      match: {
        params: {
          current_level: 3,
          direction: 'down',
          node_id: '13',
          risk_highlighter: 'high_risk',
          title: 'Total Pregnancies',
        } as any,
      },
    };
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedHierarchichalDataTable {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('tbody'))).toMatchSnapshot('hierarchichalDataTable');

    props = {
      match: {
        params: {
          current_level: 2,
          direction: 'up',
          from_level: 3,
          node_id: '13',
          risk_highlighter: 'high_risk',
          title: 'Total Pregnancies',
        } as any,
      },
    };
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedHierarchichalDataTable {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('tbody'))).toMatchSnapshot('hierarchichalDataTable');

    props = {
      match: {
        params: {
          current_level: 1,
          direction: 'up',
          from_level: 3,
          node_id: '13',
          risk_highlighter: 'high_risk',
          title: 'Total Pregnancies',
        } as any,
      },
    };
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedHierarchichalDataTable {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('tbody'))).toMatchSnapshot('hierarchichalDataTable');

    props = {
      match: {
        params: {
          current_level: 0,
          direction: 'up',
          from_level: 3,
          node_id: '13',
          risk_highlighter: 'high_risk',
          title: 'Total Pregnancies',
        } as any,
      },
    };
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedHierarchichalDataTable {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('tbody'))).toMatchSnapshot('hierarchichalDataTable');

    props = {
      match: {
        params: {
          current_level: 1,
          direction: 'up',
          from_level: 2,
          node_id: '13',
          risk_highlighter: 'high_risk',
          title: 'Total Pregnancies',
        } as any,
      },
    };
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedHierarchichalDataTable {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('tbody'))).toMatchSnapshot('hierarchichalDataTable');
    props = {
      match: {
        params: {
          current_level: 0,
          direction: 'up',
          from_level: 1,
          node_id: '13',
          risk_highlighter: 'high_risk',
          title: 'Total Pregnancies',
        } as any,
      },
    };
    wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <ConnectedHierarchichalDataTable {...props} />
        </Router>
      </Provider>
    );
    expect(toJson(wrapper.find('tbody'))).toMatchSnapshot('hierarchichalDataTable');
  });
});
