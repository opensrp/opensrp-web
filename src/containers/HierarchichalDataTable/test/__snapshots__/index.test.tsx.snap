// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchichalDataTable renders correctly: hierarchichalDataTable 1`] = `
<Table
  borderless={true}
  responsiveTag="div"
  striped={true}
  tag="table"
>
  <table
    className="table table-borderless table-striped"
  >
    <thead
      id="header"
    >
      <tr>
        <th
          className="default-width"
        />
        <th
          className="default-width"
        >
          High Risk
        </th>
        <th
          className="default-width"
        >
          Low Risk
        </th>
        <th
          className="default-width"
        >
          No Risk
        </th>
        <th
          className="default-width"
        >
          Total
        </th>
      </tr>
    </thead>
    <tbody
      id="body"
    >
      <tr
        key="d79168ed-ae95-43d1-8e45-9078212815f6"
      >
        <td
          className="default-width"
        >
          <Link
            to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/d79168ed-ae95-43d1-8e45-9078212815f6"
          >
            <a
              href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/d79168ed-ae95-43d1-8e45-9078212815f6"
              onClick={[Function]}
            >
              Lang Son Province
            </a>
          </Link>
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width"
        >
          0
        </td>
      </tr>
      <tr
        key="c13b5c95-55c3-4848-9509-b82596274d96"
      >
        <td
          className="default-width"
        >
          <Link
            to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/c13b5c95-55c3-4848-9509-b82596274d96"
          >
            <a
              href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/c13b5c95-55c3-4848-9509-b82596274d96"
              onClick={[Function]}
            >
              Dien Bien
            </a>
          </Link>
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width"
        >
          0
        </td>
      </tr>
      <tr
        key="78a12165-3c12-471f-8755-c96bac123292"
      >
        <td
          className="default-width"
        >
          <Link
            to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/78a12165-3c12-471f-8755-c96bac123292"
          >
            <a
              href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/78a12165-3c12-471f-8755-c96bac123292"
              onClick={[Function]}
            >
              Hanoi
            </a>
          </Link>
        </td>
        <td
          className="default-width "
        >
          7
        </td>
        <td
          className="default-width "
        >
          2
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width"
        >
          9
        </td>
      </tr>
      <tr
        key="f7d1d9bb-95dc-43bc-8434-0dddd1f3d374"
      >
        <td
          className="default-width"
        >
          <Link
            to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/f7d1d9bb-95dc-43bc-8434-0dddd1f3d374"
          >
            <a
              href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/f7d1d9bb-95dc-43bc-8434-0dddd1f3d374"
              onClick={[Function]}
            >
              Loc Binh
            </a>
          </Link>
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width"
        >
          0
        </td>
      </tr>
      <tr
        key="total"
      >
        <td
          className="default-width"
          id="total"
        >
          Total(
          Province
          )
        </td>
        <td
          className="default-width "
        >
          7
        </td>
        <td
          className="default-width "
        >
          2
        </td>
        <td
          className="default-width "
        >
          0
        </td>
        <td
          className="default-width"
        >
          9
        </td>
      </tr>
    </tbody>
  </table>
</Table>
`;

exports[`HierarchichalDataTable renders correctly: hierarchichalDataTable 2`] = `
<tbody
  id="body"
>
  <tr
    id="no-rows"
  >
    <td>
      There seems to be no rows here :-(
    </td>
  </tr>
  <tr
    key="total"
  >
    <td
      className="default-width"
      id="total"
    >
      Total(
      District
      )
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
</tbody>
`;

exports[`HierarchichalDataTable renders correctly: hierarchichalDataTable 3`] = `
<tbody
  id="body"
>
  <tr
    id="no-rows"
  >
    <td>
      There seems to be no rows here :-(
    </td>
  </tr>
  <tr
    key="total"
  >
    <td
      className="default-width"
      id="total"
    >
      Total(
      Commune
      )
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
</tbody>
`;

exports[`HierarchichalDataTable renders correctly: hierarchichalDataTable 4`] = `
Array [
  <tbody
    id="body"
  >
    <tr
      id="no-rows"
    >
      <td>
        There seems to be no rows here :-(
      </td>
    </tr>
    <tr
      key="total"
    >
      <td
        className="default-width"
        id="total"
      >
        Total(
        Village
        )
      </td>
      <td
        className="default-width "
      >
        0
      </td>
      <td
        className="default-width "
      >
        0
      </td>
      <td
        className="default-width "
      >
        0
      </td>
      <td
        className="default-width"
      >
        0
      </td>
    </tr>
  </tbody>,
  <tbody
    id="body"
  />,
]
`;

exports[`HierarchichalDataTable renders correctly: hierarchichalDataTable 5`] = `
<tbody
  id="body"
>
  <tr
    id="no-rows"
  >
    <td>
      There seems to be no rows here :-(
    </td>
  </tr>
  <tr
    key="total"
  >
    <td
      className="default-width"
      id="total"
    >
      Total(
      Commune
      )
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
</tbody>
`;

exports[`HierarchichalDataTable renders correctly: hierarchichalDataTable 6`] = `
<tbody
  id="body"
>
  <tr
    id="no-rows"
  >
    <td>
      There seems to be no rows here :-(
    </td>
  </tr>
  <tr
    key="total"
  >
    <td
      className="default-width"
      id="total"
    >
      Total(
      District
      )
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
</tbody>
`;

exports[`HierarchichalDataTable renders correctly: hierarchichalDataTable 7`] = `
<tbody
  id="body"
>
  <tr
    key="d79168ed-ae95-43d1-8e45-9078212815f6"
  >
    <td
      className="default-width"
    >
      <Link
        to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/d79168ed-ae95-43d1-8e45-9078212815f6"
      >
        <a
          href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/d79168ed-ae95-43d1-8e45-9078212815f6"
          onClick={[Function]}
        >
          Lang Son Province
        </a>
      </Link>
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
  <tr
    key="c13b5c95-55c3-4848-9509-b82596274d96"
  >
    <td
      className="default-width"
    >
      <Link
        to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/c13b5c95-55c3-4848-9509-b82596274d96"
      >
        <a
          href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/c13b5c95-55c3-4848-9509-b82596274d96"
          onClick={[Function]}
        >
          Dien Bien
        </a>
      </Link>
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
  <tr
    key="78a12165-3c12-471f-8755-c96bac123292"
  >
    <td
      className="default-width"
    >
      <Link
        to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/78a12165-3c12-471f-8755-c96bac123292"
      >
        <a
          href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/78a12165-3c12-471f-8755-c96bac123292"
          onClick={[Function]}
        >
          Hanoi
        </a>
      </Link>
    </td>
    <td
      className="default-width "
    >
      7
    </td>
    <td
      className="default-width "
    >
      2
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      9
    </td>
  </tr>
  <tr
    key="f7d1d9bb-95dc-43bc-8434-0dddd1f3d374"
  >
    <td
      className="default-width"
    >
      <Link
        to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/f7d1d9bb-95dc-43bc-8434-0dddd1f3d374"
      >
        <a
          href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/f7d1d9bb-95dc-43bc-8434-0dddd1f3d374"
          onClick={[Function]}
        >
          Loc Binh
        </a>
      </Link>
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
  <tr
    key="total"
  >
    <td
      className="default-width"
      id="total"
    >
      Total(
      Province
      )
    </td>
    <td
      className="default-width "
    >
      7
    </td>
    <td
      className="default-width "
    >
      2
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      9
    </td>
  </tr>
</tbody>
`;

exports[`HierarchichalDataTable renders correctly: hierarchichalDataTable 8`] = `
<tbody
  id="body"
>
  <tr
    id="no-rows"
  >
    <td>
      There seems to be no rows here :-(
    </td>
  </tr>
  <tr
    key="total"
  >
    <td
      className="default-width"
      id="total"
    >
      Total(
      District
      )
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
</tbody>
`;

exports[`HierarchichalDataTable renders correctly: hierarchichalDataTable 9`] = `
<tbody
  id="body"
>
  <tr
    key="d79168ed-ae95-43d1-8e45-9078212815f6"
  >
    <td
      className="default-width"
    >
      <Link
        to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/d79168ed-ae95-43d1-8e45-9078212815f6"
      >
        <a
          href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/d79168ed-ae95-43d1-8e45-9078212815f6"
          onClick={[Function]}
        >
          Lang Son Province
        </a>
      </Link>
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
  <tr
    key="c13b5c95-55c3-4848-9509-b82596274d96"
  >
    <td
      className="default-width"
    >
      <Link
        to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/c13b5c95-55c3-4848-9509-b82596274d96"
      >
        <a
          href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/c13b5c95-55c3-4848-9509-b82596274d96"
          onClick={[Function]}
        >
          Dien Bien
        </a>
      </Link>
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
  <tr
    key="78a12165-3c12-471f-8755-c96bac123292"
  >
    <td
      className="default-width"
    >
      <Link
        to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/78a12165-3c12-471f-8755-c96bac123292"
      >
        <a
          href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/78a12165-3c12-471f-8755-c96bac123292"
          onClick={[Function]}
        >
          Hanoi
        </a>
      </Link>
    </td>
    <td
      className="default-width "
    >
      7
    </td>
    <td
      className="default-width "
    >
      2
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      9
    </td>
  </tr>
  <tr
    key="f7d1d9bb-95dc-43bc-8434-0dddd1f3d374"
  >
    <td
      className="default-width"
    >
      <Link
        to="/hierarchicaldata/high_risk/Total Pregnancies/1/down/f7d1d9bb-95dc-43bc-8434-0dddd1f3d374"
      >
        <a
          href="/hierarchicaldata/high_risk/Total Pregnancies/1/down/f7d1d9bb-95dc-43bc-8434-0dddd1f3d374"
          onClick={[Function]}
        >
          Loc Binh
        </a>
      </Link>
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      0
    </td>
  </tr>
  <tr
    key="total"
  >
    <td
      className="default-width"
      id="total"
    >
      Total(
      Province
      )
    </td>
    <td
      className="default-width "
    >
      7
    </td>
    <td
      className="default-width "
    >
      2
    </td>
    <td
      className="default-width "
    >
      0
    </td>
    <td
      className="default-width"
    >
      9
    </td>
  </tr>
</tbody>
`;
